flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayExceptFunction.java

```diff
@@ -0,0 +1,161 @@
+
+    private transient MethodHandle equalityHandle;
+
+    public ArrayExceptFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_EXCEPT, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType()
+                        .toInternal();
+        elementGetter = ArrayData.createElementGetter(dataType.toInternal().getLogicalType());
+        hashcodeEvaluator =
+                context.createEvaluator(
+                        Expressions.call("$HASHCODE$1", $("element1")),
+                        DataTypes.INT(),
+                        DataTypes.FIELD("element1", dataType.notNull().toInternal()));
+        equalityEvaluator =
+                context.createEvaluator(
+                        $("element1").isEqual($("element2")),
+                        DataTypes.BOOLEAN(),
+                        DataTypes.FIELD("element1", dataType.notNull().toInternal()),
+                        DataTypes.FIELD("element2", dataType.notNull().toInternal()));
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        hashcodeHandle = hashcodeEvaluator.open(context);
+        equalityHandle = equalityEvaluator.open(context);
+    }
+
+    public @Nullable ArrayData eval(ArrayData arrayOne, ArrayData arrayTwo) {
+        try {
+            if (arrayOne == null) {
+                return null;
+            }
+
+            List<Object> list = new ArrayList<>();
+            Set<ObjectContainer> seen = new HashSet<>();
+
+            boolean isNullPresentInArrayTwo = false;
+            if (arrayTwo != null) {
+                for (int pos = 0; pos < arrayTwo.size(); pos++) {
+                    final Object element = elementGetter.getElementOrNull(arrayTwo, pos);
+                    if (element == null) {
+                        isNullPresentInArrayTwo = true;
+                    } else {
+                        ObjectContainer objectContainer = new ObjectContainer(element);
+                        seen.add(objectContainer);
+                    }
+                }
+            }
+            boolean isNullPresentInArrayOne = false;
+            for (int pos = 0; pos < arrayOne.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayOne, pos);
+                if (element == null) {
+                    isNullPresentInArrayOne = true;
+                } else {
+                    ObjectContainer objectContainer = new ObjectContainer(element);
+                    if (!seen.contains(objectContainer)) {
+                        seen.add(objectContainer);
+                        list.add(element);
+                    }
+                }
+            }
+            if (!isNullPresentInArrayTwo && isNullPresentInArrayOne) {
+                list.add(null);
+            }
```

Comment:
Is this what we want? Have you checked this is what other DBs do? If that's the case can we add this behaviour to the docs?

Personally, I find it really surprising.

Reply:
Ok I have already done some research, here is what I find in others DBs.
For snowflake:   [](https://docs.snowflake.com/en/sql-reference/functions/array_except)
```
Returns
This function returns an ARRAY that contains the elements from source_array that are not in array_of_elements_to_exclude.
If no elements remain after excluding the elements in array_of_elements_to_exclude from source_array, the function returns an empty ARRAY.
If one or both arguments are NULL, the function returns NULL.
The order of the values within the returned array is unspecified.
```
So for snowflake the the order of the values within the returned array is unspecified.

for Databrick: [](https://docs.databricks.com/en/sql/language-manual/functions/array_except.html)
it also not explain the null situation and the order of the values within the returned array

for spark: [](https://spark.apache.org/docs/3.1.3/api/python/reference/api/pyspark.sql.functions.array_except.html)
it also not explain the null situation and the order of the values within the returned array

for prestodb:   [](https://prestodb.io/docs/current/search.html?q=array_except#)
it also not explain the null situation and the order of the values within the returned array

for Doris: [](https://doris.apache.org/docs/1.2/sql-manual/sql-functions/array-functions/array_except/)
The order of order of the values within the returned array is specified. the order is same as array1
But if any input array is null, it will return null in the end.

Reply:
I see there is one more question to answear from Snowflake's docs:

> ARRAY_EXCEPT compares arrays by using multi-set semantics (sometimes called “bag semantics”). If source_array includes multiple copies of a value, the function only removes the number of copies of that value that are specified in array_of_elements_to_exclude.

> In other words, if source_array has N copies of a value and array_of_elements_to_exclude has M copies of the same value, the function excludes M copies of the value from the returned array. The number of copies of the value in the returned array is N - M or, if M is larger than N, 0.

> For example, if source_array contains 5 elements with the value 'A' and array_of_elements_to_exclude contains 2 elements with the value 'A', the returned array contains 3 elements with the value 'A'.

Do we want that behaviour or deduplicate behaviour?

cc @MartijnVisser WDYT?

1. Should we deduplicate (set semantics) or remove only the number of elements from array2 (multi-set) ?
2. What should be the order in the resulting array?

Reply:
btw I believe @snuyanzin suggested not removing duplicates here: https://github.com/apache/flink/pull/22629#issuecomment-1577379764

Reply:
Snowflake behaves like this:

```sql
SELECT ARRAY_EXCEPT(['A', 'A', 'C', 'C', NULL, NULL, 'B'], ['A']);
```

Returns:
```sql
[
  "A",
  "C",
  "C",
  NULL,
  NULL,
  "B"
]
```

Databricks / Spark behaves like this:

```sql
SELECT array_except(array('A', 'A', 'C', 'C', NULL, NULL, 'B'), array('A')) AS result
```

Returns:
```sql
["C", null, "B"]
```

So from my point of view:
1. We should not move `null` the end, but keep the existing place in the array
2. We should not remove duplicates: if the user wants to remove duplicates, they can run an `ARRAY_DISTINCT` like @snuyanzin also pointed out

So bottom line I think we should follow the Snowflake implementation

Reply:
@MartijnVisser @dawidwys 
If we adopt the Snowflake implementation, the order of values in the returned array is unspecified, according to Snowflake documentation (https://docs.snowflake.com/en/sql-reference/functions/array_except). Therefore, do we not need to concern ourselves with the order of the values, similar to Snowflake's approach?

Reply:
I'm not sure how much guarantees we can provide to users. Ideally you would like to guarantee ordering, but that's a trade-off against effort of course.

Reply:
I think we don't need to copy 1:1 behavior
we could take something like ability to keep duplicates since it will allow cover more cases (`array_distinct` could cover cases w/o). 
Also in some cases IMHO it would be useful to keep the original order, so the question here: is there anything blocking or making it very difficult to maintain it here?

Reply:
ok